elastic/kibana
# [Monitoring][Alerting] CCR read exceptions alert

Resolves: https://github.com/elastic/kibana/issues/79990

The alerting query groups remote clusters with their relative follow indices. This isn't a per node type an alert nor is it a threshold type of alert. It can either be enabled or disabled.

UI/UX:
![Screen Shot 2020-12-15 at 12 39 02 AM](https://user-images.githubusercontent.com/27822304/102175766-15eb5100-3e6e-11eb-8695-a27b82cbe7b7.png)

![Screen Shot 2020-12-15 at 12 24 51 AM](https://user-images.githubusercontent.com/27822304/102175783-1be13200-3e6e-11eb-93b6-861d98205073.png)

![Screen Shot 2020-12-15 at 12 24 59 AM](https://user-images.githubusercontent.com/27822304/102175793-20a5e600-3e6e-11eb-9584-4fb1923e34f7.png)

![Screen Shot 2020-12-15 at 12 25 09 AM](https://user-images.githubusercontent.com/27822304/102175839-30252f00-3e6e-11eb-93e1-d4604092772a.png)

![Screen Shot 2020-12-15 at 12 25 32 AM](https://user-images.githubusercontent.com/27822304/102175854-36b3a680-3e6e-11eb-80f7-d3e690baa521.png)

---

**Testing:**

1. Setup a basic CCR environment, [doc link](https://www.elastic.co/guide/en/elasticsearch/reference/current/ccr-getting-started.html) | [x-pack ver](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/xpack-ccr.html)

2. Confirm you have a CCR setup with a leader/follower relationship via Kibana:
`.../app/management/data/cross_cluster_replication` and that the [CRR monitoring metrics](https://www.elastic.co/guide/en/kibana/current/elasticsearch-metrics.html#ccr-overview-page) are working

3. I don't know a "legitimate" way of triggering the read exceptions, so I'm doing it manually, eg:
```JSON
PUT .monitoring-es-7-2020.12.15/_doc/abc123
{
  "cluster_uuid": "BcK-0pmsQniyPQfZuauuXw",
  "timestamp": "2020-12-15T04:36:44.402Z",
  "interval_ms": 10000,
  "type": "ccr_stats",
  "source_node": {
    "uuid": "WJaWz2XIR8mqsZDy7eeykA",
    "host": "10.46.8.145",
    "transport_address": "10.46.8.145:19157",
    "ip": "10.46.8.145",
    "name": "instance-0000000000",
    "timestamp": "2020-12-10T07:01:52.391Z"
  },
  "ccr_stats": {
    "remote_cluster": "BcK-0pmsQniyPQfZuauuXw_remote_cluster_1",
    "leader_index": ".leader_index_1",
    "follower_index": ".follower_index_1",
    "shard_id": 1,
    "leader_global_checkpoint": 2,
    "leader_max_seq_no": 3,
    "follower_global_checkpoint": 4,
    "follower_max_seq_no": 5,
    "last_requested_seq_no": 6,
    "outstanding_read_requests": 7,
    "outstanding_write_requests": 8,
    "write_buffer_operation_count": 9,
    "write_buffer_size_in_bytes": 10,
    "follower_mapping_version": 11,
    "follower_settings_version": 12,
    "follower_aliases_version": 13,
    "total_read_time_millis": 14,
    "total_read_remote_exec_time_millis": 15,
    "read_exceptions": [
      {
        "exception": {
          "type": "read_exceptions_type_1",
          "reason": "read_exceptions_reason_1"
        }
      }
    ],
    "successful_read_requests": 1,
    "failed_read_requests": 1,
    "operations_read": 3,
    "bytes_read": 4,
    "total_write_time_millis": 5,
    "successful_write_requests": 6,
    "failed_write_requests": 7,
    "operations_written": 8,
    "time_since_last_read_millis": 9,
    "fatal_exception": {
      "type": "fatal_exception_type",
      "reason": "fatal_exception_reason"
    }
  }
}
```
Be sure to update the `timestamp` and the index name to be more recent
